README.md
# express-tracking

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![Total Downloads][ico-downloads]][link-downloads]

express tracking 快递单号状态追踪

## Install

Via Composer

``` bash
$ composer require xu42/express-tracking
```

## Usage

Please see [debug.php](debug.php)

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Testing

Tests unavailable.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Security

If you discover any security related issues, please using the issue tracker.

## Credits

- [Xu42](https://github.com/xu42)
- [All Contributors](https://github.com/xu42/express-tracking/contributors)

## License

The GPL2.0 License. Please see [License File](LICENSE.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/xu42/express-tracking.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/xu42/express-tracking.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/xu42/express-tracking
[link-downloads]: https://packagist.org/packages/xu42/express-tracking
[link-author]: https://github.com/xu42
[link-contributors]: ../../contributors


CHANGELOG.md
# Changelog

## 2.0.0 - 2017-04-23

* 大版本更新, 添加异常抛出, 接口地址改变

## 1.0.6 - 2016-09-11

* 修复申通被错误识别为中通的Bug (完善)

## 1.0.0 - 2016-06-11

* First release


src/BizImpl/ExpressTrackingBizImpl.php
<?php

namespace Cn\Xu42\ExpressTracking\BizImpl;

use Cn\Xu42\ExpressTracking\Exception\ArgumentException;
use Cn\Xu42\ExpressTracking\Exception\SystemException;

class ExpressTrackingBizImpl
{
    const URL_WAP = 'http://m.kuaidi100.com/result.jsp?from=weixin&nu=';
    const URL_COM_CODE = 'http://m.kuaidi100.com/autonumber/auto?num=';
    const URL_QUERY = 'http://m.kuaidi100.com/query?';


    public function getComCodes($postId)
    {
        if (empty($postId)) throw new ArgumentException('运单号格式错误');

        $url = self::URL_COM_CODE . $postId;
        $curlResponse = $this->curlRequest($url);
        $comCodes = [];

        foreach (json_decode($curlResponse, true) as $item) {
            $comCodes[] = isset($item['comCode']) ? $item['comCode'] : null;
        }

        if (empty($comCodes)) throw new SystemException('comCode查询不到');

        return $comCodes;
    }


    public function query($postId, $comCodes)
    {
        if (empty($postId)) throw new ArgumentException('运单号格式错误');

        if (empty($comCodes) || !is_array($comCodes)) throw new ArgumentException('comCodes格式错误');

        $result = [];
        foreach ($comCodes as $comCode) {
            $result = $this->getQueryResult($comCode, $postId);
            if ($result['status'] == '200') break;
        }

        if (empty($result) || $result['message'] != 'ok') throw new SystemException('查询不到');

        $result['url'] = self::URL_WAP . $postId;

        return $result;
    }


    private function curlRequest($url)
    {
        $ch = curl_init();
        curl_setopt($ch, CURLOPT_URL, $url);
        curl_setopt($ch, CURLOPT_REFERER, $url);
        curl_setopt($ch, CURLOPT_TIMEOUT, 5);
        curl_setopt($ch, CURLOPT_BINARYTRANSFER, true);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_USERAGENT, 'Mozilla/5.0 (Windows NT 6.1; WOW64; rv:24.0) Gecko/20100101 Firefox/24.0');
        $curlResponse = curl_exec($ch);
        curl_close($ch);
        if ($curlResponse === false) throw new SystemException('网络请求失败');
        return $curlResponse;
    }


    private function getQueryResult($comCode, $postId)
    {
        $url = self::URL_QUERY . 'type=' . $comCode . '&postid=' . $postId . '&id=1&valicode=&temp=' . time();
        $response = $this->curlRequest($url);
        return json_decode($response, true);
    }

}

src/Service/ExpressTrackingService.php
<?php

namespace Cn\Xu42\ExpressTracking\Service;

use Cn\Xu42\ExpressTracking\Exception\SystemException;
use Cn\Xu42\ExpressTracking\BizImpl\ExpressTrackingBizImpl;

class ExpressTrackingService
{
    private $bizImpl = null;

    public function __construct()
    {
        $this->bizImpl = new ExpressTrackingBizImpl();
    }

    /**
     * 获取快递运单号可能所属的公司
     *
     * @param $postId
     * @return array
     * @throws SystemException
     */
    public function getComCodes($postId)
    {
        try {
            return $this->bizImpl->getComCodes($postId);
        } catch (\Throwable $throwable) {
            throw new SystemException($throwable->getMessage());
        }
    }

    /**
     * 获取快递运单号最新动态
     *
     * @param $postId
     * @param $comCodes
     * @return array|mixed
     * @throws SystemException
     */
    public function query($postId, $comCodes)
    {
        try {
            return $this->bizImpl->query($postId, $comCodes);
        } catch (\Throwable $throwable) {
            throw new SystemException($throwable->getMessage());
        }
    }

}


src/ExpressTracking.php
<?php

namespace Xu42\ExpressTracking;

/**
 * 快递单号状态追踪
 * Class ExpressTracking
 * @package Xu42\ExpressTracking
 */
class ExpressTracking
{
    private $num = '';
//    private $comCode    = '';
    private $urlWap     = 'http://m.kuaidi100.com/result.jsp?from=weixin&nu=';
    private $urlComCode = 'http://m.kuaidi100.com/autonumber/auto?num=';
    private $urlQuery   = 'http://m.kuaidi100.com/query?';

    /**
     * expressTracking constructor.
     * @param string $webUrl
     */
    public function __construct( $num )
    {
        $this->num        = $num;
        $this->urlWap     = $this->urlWap . $num;
        $this->urlComCode = $this->urlComCode . $num;
    }

    /**
     * 获取一快递单号的最新动态
     * status 200表示查询正常, 其它为非异常结果
     * @return mixed
     */
    public function latestStatus()
    {
        $result = $this->query();
        if ( is_null( $result ) ) {
            return ['status' => '404', 'message' => '未查询到快递状态', 'data' => [], 'url' => $this->urlWap];
        }
        if ( $result['message'] == 'ok' ) {
            $result['url'] = $this->urlWap;
            return $result;
        }
        return ['status' => $result['status'], 'message' => $result['message'], 'data' => [], 'url' => $this->urlWap];
    }

    /**
     * 简单封装的get网络请求助手
     * @param $url
     * @return mixed
     */
    private function myCurl( $url )
    {
        $ch = curl_init();
        curl_setopt( $ch, CURLOPT_URL, $url );
        curl_setopt( $ch, CURLOPT_REFERER, $this->urlWap );
        curl_setopt( $ch, CURLOPT_RETURNTRANSFER, true );
        curl_setopt( $ch, CURLOPT_CONNECTTIMEOUT, 5 );
        curl_setopt( $ch, CURLOPT_USERAGENT,
            'Mozilla/5.0 (Windows NT 6.1; WOW64; rv:24.0) Gecko/20100101 Firefox/24.0' );
        $webPage = curl_exec( $ch );
        curl_close( $ch );
        return $webPage;
    }

    /**
     * 获取快递单号所属的快递公司代号
     * @return mixed
     */
    private function getComCode()
    {
        $response = $this->myCurl( $this->urlComCode );
        $comCode  = json_decode( $response, true );
        $com      = [];
        if ( isset($comCode[0]) ) {
//            $this->comCode = $comCode[0]['comCode'];
//            return $comCode[0]['comCode'];
            $com[] = $comCode[0]['comCode'];
            if ( isset($comCode[1]) ) {
                $com[] = $comCode[1]['comCode'];
            }
            return $com;
        }
        return null;
    }

    /**
     * 查询快递状态
     * @return mixed
     */
    private function query()
    {

        $type = $this->getComCode();

        if ( is_null( $type ) ) {
            return null;
        }
        $result = $this->getQueryResult( $type[0], $this->num );
        if ($result['status'] == '200'){
            return $result;
        }

        if ( $result['status'] == '201' && count( $type ) >= 2 ) {
            $result1 = $this->getQueryResult( $type[1], $this->num );
            if ($result1['status'] == '200') {
                return $result1;
            }
        }
        return $result;
    }

    /**
     * 获取查询结果
     * @param $type
     * @param $postId
     * @return mixed
     */
    private function getQueryResult( $type, $postId )
    {
        $url      = $this->urlQuery . 'type=' . $type . '&postid=' . $postId . '&id=1&valicode=&temp=' . time();
        $response = $this->myCurl( $url );
        $result   = json_decode( $response, true );
        return $result;
    }
}
